Schedule counts balancing-related partition movements in partition balancer

ztlpn opened the pull request:
Move counts balancing code from `members_backend` to the partition balancer.

The algorithm is as follows:
* Maintain a list of nodes for which `reallocations_finished` command hasn't been issued yet. If this list is not empty, run the balancing operator in the partition balancer planner
* Planner simply goes over all replicas and tries to move them to a more optimal place.
* If we haven't added any actions - this means we are at a (local) minimum and rebalancing is finished => we can issue `reallocations_finished` commands.

On-demand rebalancing is moved to the balancer as well.

As during counts-based rebalancing sometimes moves are generated that do not result in improvement, revert API is implemented to discard them.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes
### Improvements
* Partition movements related to rebalancing on node addition now take free disk space and health of target nodes into account.

mmaslankaprv commented:
this looks really good, i am wondering if we can remove update finishing logic from the members backend now ?

ztlpn commented:
> i am wondering if we can remove update finishing logic from the members backend now ?

We need to get rid of the recommission-related cancellations logic first.

ztlpn commented:
Rebased on dev and added [another change](https://github.com/redpanda-data/redpanda/compare/3fba0c5d5467ad044d44dde77089abf6c5f5329a..d6ea177c0d5e425567aeab1459f2628db6689e2c): tightened the way we handle per-term state in the backend. This way it is easier to remember to reset it when the term changes.

ztlpn commented:
Change in [force-push](https://github.com/redpanda-data/redpanda/compare/c45d4f0dce5fcdae88058b65cdf813f63ffbaec7..17045c146033bdcfb328532ec9f370f7b1a5be0a): added controller term check when issuing finish_reallocations commands.

ztlpn commented:
[test failures](https://buildkite.com/redpanda/redpanda/builds/31351)

* https://github.com/redpanda-data/redpanda/issues/11304
* https://github.com/redpanda-data/redpanda/issues/11044
* https://github.com/redpanda-data/redpanda/issues/11276
* https://github.com/redpanda-data/redpanda/issues/11456
* https://github.com/redpanda-data/redpanda/issues/10497

ztlpn commented:
/ci-repeat 2
skip-units
dt-repeat=10
tests/rptest/tests/nodes_decommissioning_test.py
tests/rptest/tests/scaling_up_test.py
tests/rptest/tests/partition_balancer_test.py
tests/rptest/tests/random_node_operations_test.py

ztlpn commented:
/ci-repeat 2
skip-units
dt-repeat=10
tests/rptest/tests/nodes_decommissioning_test.py
tests/rptest/tests/scaling_up_test.py
tests/rptest/tests/partition_balancer_test.py
tests/rptest/tests/random_node_operations_test.py
